src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/message_infrastructure_logging.h

```diff
@@ -5,39 +5,157 @@
+#if DEBUG_MODE && DEBUG_LOG_TO_FILE
+#define DEBUG_LOG_PRINT(_level,_module, _fmt, ...) do {\
+  int length = 0;\
+  char * log_data = (char*)malloc(sizeof(char)*MAX_SIZE_PER_LOG_MSG);\
+  if (log_data != NULL)\
+    length = std::snprintf(log_data, MAX_SIZE_PER_LOG_MSG, _fmt,  ## __VA_ARGS__);\
+  if (log_data == NULL || length <0){\
+    getLogInstance()->log(new LogMsg(std::string(LOG_MSG_SUBSTITUTION), __FILE__, __LINE__, _module, _level));\
+  }else{\
+    getLogInstance()->log(new LogMsg(std::string(log_data), __FILE__, __LINE__, _module, _level));\
+  }\
+  free(log_data);\
+}while(0)
+
+#else
+#define DEBUG_LOG_PRINT(_level,_module, ...) //std::printf(__VA_ARGS__)
+#endif
+
+#define LAVA_LOG(_module, _fmt, ...) { \
+  if ((_module)) { \
+    DEBUG_LOG_PRINT("[CPP INFO]", _module, _fmt, ## __VA_ARGS__);\
   } \
 }
 
-#define LAVA_DUMP(_cond, _fmt, ...) { \
-  if ((_cond && DEBUG_MODE)) { \
-    printf(_fmt, ## __VA_ARGS__); \
+#define LAVA_DUMP(_module, _fmt, ...) { \
+  if ((_module && DEBUG_MODE)) { \
+    DEBUG_LOG_PRINT("[CPP DUMP]", _module, _fmt, ## __VA_ARGS__);\
   } \
 }
 
-#define LAVA_DEBUG(_cond, _fmt, ...) { \
-  if ((_cond && DEBUG_MODE)) { \
-    printf("[CPP DEBUG]" _fmt, ## __VA_ARGS__); \
+#define LAVA_DEBUG(_module, _fmt, ...) { \
+  if ((_module && DEBUG_MODE)) { \
+    DEBUG_LOG_PRINT("[CPP DBUG]", _module, _fmt, ## __VA_ARGS__);\
   } \
 }
 
-#define LAVA_LOG_WARN(_cond, _fmt, ...) { \
-  if ((_cond)) { \
-    printf("[CPP WARNING] %s " _fmt, __FUNCTION__, ## __VA_ARGS__); \
+#define LAVA_LOG_WARN(_module, _fmt, ...) { \
+  if ((_module)) { \
+    DEBUG_LOG_PRINT("[CPP WARN]", _module, _fmt, ## __VA_ARGS__);\
   } \
 }
 
-#define LAVA_LOG_ERR(_fmt, ...) { \
-  printf("[CPP ERROR] %s " _fmt, __FUNCTION__, ## __VA_ARGS__); \
+#define LAVA_LOG_ERR(_module, _fmt, ...) { \
+  DEBUG_LOG_PRINT("[CPP ERRO]", _module, _fmt, ## __VA_ARGS__);\
 }
 
-#endif  // MESSAGE_INFRASTRUCTURE_LOGGING_H_
+namespace message_infrastructure {
+
+enum Log_Level{
+  DISABLE_LOG,
+  ERROR,
+  WARN,
+  LOG,
+  DUMP,
+  DEBUG
+};
+
+enum Log_Module{
+  LOG_NULL,
+  LOG_MP,       // log for multiprocessing
+  LOG_ACTOR,
+  LOG_LAYER,
+  LOG_SMMP      // log for shmemport
+};
+
+class LogMsg{
+  public:
+    LogMsg(const std::string&, const char* log_file, int log_line, enum Log_Module log_module, const char* log_level);
+    std::string getEntireLogMsg(int pid);
+
+  private:
+  // std::string msg_time_ = nullptr; this way will cause a abort trouble...
+    std::string msg_time_;
+    std::string msg_data_;
+    std::string msg_level_;
+    std::string msg_module_;
+    std::string msg_file_ ;
+    std::string msg_pid_;
+    int msg_line_ = 0;
+};
+
+class Log{
+  public:
+    Log();
+    ~Log();
+    void log(LogMsg*);
+    void clear();
+    void write_down();
+    
+
+  private:
+    std::string log_path_;
+    int current_length_ = 0;
+    const int max_length_ = MAX_SIZE_LOG;
+    std::mutex log_lock_;
+    LogMsg * log_buffer_[MAX_SIZE_LOG];
```

Comment:
use `std::queue` to replace logic about `current_length_`, `max_length_`, `log_buffer_`